File: packages/react-native/ios/Capture/RJPerformanceManager.m
//
//  RJPerformanceManager.m
//  Rejourney
//
//  System performance monitoring and adaptive throttling implementation.
//
//  Copyright (c) 2026 Rejourney
//

#import "RJPerformanceManager.h"
#import "../Core/RJConstants.h"
#import "../Core/RJLogger.h"
#import <UIKit/UIKit.h>
#import <mach/mach.h>

#pragma mark - Private Interface

@interface RJPerformanceManager ()

@property(nonatomic, assign) RJPerformanceLevel currentLevel;

@property(nonatomic, strong, nullable) dispatch_source_t memoryPressureSource;

@property(nonatomic, strong, nullable) NSTimer *cpuMonitorTimer;

@property(nonatomic, assign) float rollingCPUAverage;

@property(nonatomic, assign) NSInteger highCPUSampleCount;

@property(nonatomic, assign) BOOL isMonitoring;

@end

#pragma mark - Implementation

@implementation RJPerformanceManager

#pragma mark - Singleton

+ (instancetype)sharedManager {
  static RJPerformanceManager *sharedInstance = nil;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    sharedInstance = [[self alloc] init];
  });
  return sharedInstance;
}

#pragma mark - Initialization

- (instancetype)init {
  self = [super init];
  if (self) {
    _currentLevel = RJPerformanceLevelNormal;
    _thermalThrottleEnabled = YES;
    _batteryAwareEnabled = YES;
    _isMonitoring = NO;
    _rollingCPUAverage = 0.0;
    _highCPUSampleCount = 0;
  }
  return self;
}

- (void)dealloc {
  [self stopMonitoring];
}

#pragma mark - Monitoring Lifecycle

- (void)startMonitoring {
  if (self.isMonitoring)
    return;

  self.isMonitoring = YES;

  [self setupMemoryPressureMonitoring];

  [[NSNotificationCenter defaultCenter]
      addObserver:self
         selector:@selector(handleMemoryWarningNotification:)
             name:UIApplicationDidReceiveMemoryWarningNotification
           object:nil];

  if (@available(iOS 11.0, *)) {
    [[NSNotificationCenter defaultCenter]
        addObserver:self
           selector:@selector(handleThermalStateChange:)
               name:NSProcessInfoThermalStateDidChangeNotification
             object:nil];
  }

  [[UIDevice currentDevice] setBatteryMonitoringEnabled:YES];
  [[NSNotificationCenter defaultCenter]
      addObserver:self
         selector:@selector(handleBatteryStateChange:)
             name:UIDeviceBatteryLevelDidChangeNotification
           object:nil];

  __weak typeof(self) weakSelf = self;
  self.cpuMonitorTimer =
      [NSTimer scheduledTimerWithTimeInterval:2.0
                                      repeats:YES
                                        block:^(NSTimer *timer) {
                                          [weakSelf checkCPUUsage];
                                        }];
  [[NSRunLoop currentRunLoop] addTimer:self.cpuMonitorTimer
                               forMode:NSRunLoopCommonModes];

  RJLogDebug(@"Performance monitoring started");
}

- (void)stopMonitoring {
  if (!self.isMonitoring)
    return;

  self.isMonitoring = NO;

  [[NSNotificationCenter defaultCenter] removeObserver:self];

  if (self.cpuMonitorTimer) {
    [self.cpuMonitorTimer invalidate];
    self.cpuMonitorTimer = nil;
  }

  if (self.memoryPressureSource) {
    @try {
      dispatch_source_cancel(self.memoryPressureSource);
    } @catch (NSException *e) {
    }
    self.memoryPressureSource = nil;
  }

  RJLogDebug(@"Performance monitoring stopped");
}

#pragma mark - Memory Pressure Monitoring

- (void)setupMemoryPressureMonitoring {
  dispatch_source_t source = dispatch_source_create(
      DISPATCH_SOURCE_TYPE_MEMORYPRESSURE, 0,
      DISPATCH_MEMORYPRESSURE_WARN | DISPATCH_MEMORYPRESSURE_CRITICAL,
      dispatch_get_main_queue());

  if (!source) {
    RJLogWarning(@"Failed to create memory pressure source");
    return;
  }

  __weak typeof(self) weakSelf = self;
  dispatch_source_set_event_handler(source, ^{
    __strong typeof(weakSelf) strongSelf = weakSelf;
    if (!strongSelf)
      return;

    dispatch_source_memorypressure_flags_t flags =
        dispatch_source_get_data(source);

    if (flags & DISPATCH_MEMORYPRESSURE_CRITICAL) {
      RJLogWarning(@"CRITICAL memory pressure - pausing captures");
      [strongSelf setLevel:RJPerformanceLevelPaused];
      [strongSelf handleMemoryWarning];
    } else if (flags & DISPATCH_MEMORYPRESSURE_WARN) {
      RJLogWarning(@"Memory pressure warning - reducing captures");
      [strongSelf setLevel:RJPerformanceLevelMinimal];
    }
  });

  dispatch_resume(source);
  self.memoryPressureSource = source;
}

#pragma mark - Level Management

- (void)setLevel:(RJPerformanceLevel)level {
  if (self.currentLevel != level) {
    self.currentLevel = level;
    if ([self.delegate
            respondsToSelector:@selector(performanceManagerDidChangeLevel:)]) {
      [self.delegate performanceManagerDidChangeLevel:level];
    }
  }
}

- (void)updatePerformanceLevel {
  if (!self.thermalThrottleEnabled && !self.batteryAwareEnabled) {
    [self setLevel:RJPerformanceLevelNormal];
    return;
  }

  RJPerformanceLevel newLevel = RJPerformanceLevelNormal;

  if (@available(iOS 11.0, *)) {
    if (self.thermalThrottleEnabled) {
      NSProcessInfoThermalState thermalState =
          [[NSProcessInfo processInfo] thermalState];

      switch (thermalState) {
      case NSProcessInfoThermalStateCritical:
        [self setLevel:RJPerformanceLevelPaused];
        return;
      case NSProcessInfoThermalStateSerious:
        [self setLevel:RJPerformanceLevelMinimal];
        return;
      case NSProcessInfoThermalStateFair:
        newLevel = RJPerformanceLevelReduced;
        break;
      case NSProcessInfoThermalStateNominal:
        break;
      }
    }
  }

  if (self.batteryAwareEnabled) {
    float batteryLevel = [[UIDevice currentDevice] batteryLevel];
    UIDeviceBatteryState batteryState = [[UIDevice currentDevice] batteryState];

    BOOL isCharging = (batteryState == UIDeviceBatteryStateCharging ||
                       batteryState == UIDeviceBatteryStateFull);

    if (!isCharging && batteryLevel >= 0 &&
        batteryLevel < RJLowBatteryThreshold) {
      newLevel = MAX(newLevel, RJPerformanceLevelReduced);
    }
  }

  NSUInteger usedMemory = [self currentMemoryUsage];
  if (usedMemory > RJMemoryWarningThresholdBytes) {
    newLevel = MAX(newLevel, RJPerformanceLevelReduced);
  }

  [self setLevel:newLevel];
}

#pragma mark - Notification Handlers

- (void)handleThermalStateChange:(NSNotification *)notification {
  [self updatePerformanceLevel];
}

- (void)handleBatteryStateChange:(NSNotification *)notification {
  [self updatePerformanceLevel];
}

- (void)handleMemoryWarningNotification:(NSNotification *)notification {
  [self handleMemoryWarning];
}

#pragma mark - Memory Management

- (void)handleMemoryWarning {
  RJLogWarning(@"Memory warning received");

  [self setLevel:RJPerformanceLevelMinimal];

  if ([self.delegate respondsToSelector:@selector
                     (performanceManagerDidReceiveMemoryWarning)]) {
    [self.delegate performanceManagerDidReceiveMemoryWarning];
  }
}

- (NSUInteger)currentMemoryUsage {
  struct task_basic_info info;
  mach_msg_type_number_t size = TASK_BASIC_INFO_COUNT;

  kern_return_t result =
      task_info(mach_task_self(), TASK_BASIC_INFO, (task_info_t)&info, &size);

  return (result == KERN_SUCCESS) ? info.resident_size : 0;
}

#pragma mark - CPU Monitoring

- (float)currentCPUUsage {
  thread_array_t threadList;
  mach_msg_type_number_t threadCount;

  kern_return_t kr = task_threads(mach_task_self(), &threadList, &threadCount);
  if (kr != KERN_SUCCESS) {
    return -1.0;
  }

  float totalUsage = 0.0;

  for (mach_msg_type_number_t i = 0; i < threadCount; i++) {
    thread_info_data_t threadInfo;
    mach_msg_type_number_t threadInfoCount = THREAD_INFO_MAX;

    kr = thread_info(threadList[i], THREAD_BASIC_INFO,
                     (thread_info_t)threadInfo, &threadInfoCount);

    if (kr == KERN_SUCCESS) {
      thread_basic_info_t basicInfo = (thread_basic_info_t)threadInfo;

      if (!(basicInfo->flags & TH_FLAGS_IDLE)) {
        totalUsage += basicInfo->cpu_usage / (float)TH_USAGE_SCALE * 100.0;
      }
    }

    mach_port_deallocate(mach_task_self(), threadList[i]);
  }

  vm_deallocate(mach_task_self(), (vm_address_t)threadList,
                threadCount * sizeof(thread_t));

  return totalUsage;
}

- (BOOL)isCPUHigh {
  return self.rollingCPUAverage >= 60.0;
}

- (void)checkCPUUsage {
  float currentCPU = [self currentCPUUsage];
  if (currentCPU < 0)
    return;

  if (self.rollingCPUAverage == 0) {
    self.rollingCPUAverage = currentCPU;
  } else {
    self.rollingCPUAverage = 0.3 * currentCPU + 0.7 * self.rollingCPUAverage;
  }

  /*
   * Updated CPU Thresholds:
   * CRITICAL: > 90%
   * HIGH:     > 60%
   * NORMAL:   < 40%
   */
  const float CPU_CRITICAL_THRESHOLD = 90.0; // was 80
  const float CPU_HIGH_THRESHOLD = 60.0;     // was 60
  const float CPU_NORMAL_THRESHOLD = 40.0;
  const NSInteger HYSTERESIS_SAMPLES = 3;

  RJPerformanceLevel suggestedLevel = RJPerformanceLevelNormal;

  if (self.rollingCPUAverage >= CPU_CRITICAL_THRESHOLD) {
    suggestedLevel = RJPerformanceLevelMinimal;
    self.highCPUSampleCount++;
  } else if (self.rollingCPUAverage >= CPU_HIGH_THRESHOLD) {
    // Force Reduced level (scales down to 0.4) when CPU is high
    suggestedLevel = RJPerformanceLevelReduced;
    self.highCPUSampleCount++;
  } else if (self.rollingCPUAverage < CPU_NORMAL_THRESHOLD) {
    self.highCPUSampleCount = 0;
    suggestedLevel = RJPerformanceLevelNormal;
  } else {
    // In hysteresis zone, maintain current level if it's already
    // reduced/minimal
    if (self.currentLevel >= RJPerformanceLevelReduced) {
      return;
    }
    // Otherwise keep monitoring
    return;
  }

  // Apply with hysteresis
  if (suggestedLevel > self.currentLevel) {
    if (self.highCPUSampleCount >= HYSTERESIS_SAMPLES) {
      RJLogInfo(@"CPU usage high (%.1f%%), throttling to level %ld",
                self.rollingCPUAverage, (long)suggestedLevel);
      [self setLevel:suggestedLevel];
    }
  } else if (suggestedLevel < self.currentLevel) {
    // Immediate recovery if CPU drops comfortably
    if (self.highCPUSampleCount == 0) {
      RJLogInfo(@"CPU usage normalized (%.1f%%), restoring level %ld",
                self.rollingCPUAverage, (long)suggestedLevel);
      [self setLevel:suggestedLevel];
    }
  }
}

@end

File: packages/react-native/ios/Capture/RJPerformanceManager.swift

import Foundation
import UIKit
import MachO

@objc public protocol RJPerformanceManagerDelegate: NSObjectProtocol {
    @objc optional func performanceManagerDidChangeLevel(_ level: RJPerformanceLevel)
    @objc optional func performanceManagerDidReceiveMemoryWarning()
}

@objc(RJPerformanceManager)
public class RJPerformanceManager: NSObject {
    
    @objc public static let sharedManager = RJPerformanceManager()
    
    @objc public weak var delegate: RJPerformanceManagerDelegate?
    @objc public var thermalThrottleEnabled = true
    @objc public var batteryAwareEnabled = true
    
    @objc public private(set) var currentLevel: RJPerformanceLevel = .normal {
        didSet {
            if oldValue != currentLevel {
                delegate?.performanceManagerDidChangeLevel?(currentLevel)
            }
        }
    }
    
    private var isMonitoring = false
    private var memoryPressureSource: DispatchSourceMemoryPressure?
    private var monitorTask: Task<Void, Never>?
    
    private var rollingCPUAverage: Float = 0.0
    private var highCPUSampleCount = 0
    
    // CPU Thresholds
    private let cpuCriticalThreshold: Float = 90.0
    private let cpuHighThreshold: Float = 60.0
    private let cpuNormalThreshold: Float = 40.0
    private let hysteresisSamples = 3
    
    // Low Battery Threshold (20%)
    private let lowBatteryThreshold: Float = 0.20
    
    // Memory Warning Threshold (200MB resident)
    private let memoryWarningThresholdBytes: UInt64 = 200 * 1024 * 1024
    
    private override init() {
        super.init()
    }
    
    deinit {
        stopMonitoring()
    }
    
    @objc public func startMonitoring() {
        if isMonitoring { return }
        isMonitoring = true
        
        setupMemoryPressureMonitoring()
        
        NotificationCenter.default.addObserver(self, selector: #selector(handleMemoryWarningNotification), name: UIApplication.didReceiveMemoryWarningNotification, object: nil)
        
        NotificationCenter.default.addObserver(self, selector: #selector(handleThermalStateChange), name: ProcessInfo.thermalStateDidChangeNotification, object: nil)
        
        UIDevice.current.isBatteryMonitoringEnabled = true
        NotificationCenter.default.addObserver(self, selector: #selector(handleBatteryStateChange), name: UIDevice.batteryLevelDidChangeNotification, object: nil)
        
        // Start background monitoring loop
        monitorTask = Task.detached(priority: .utility) { [weak self] in
            while !Task.isCancelled {
                // Sleep for 2 seconds
                try? await Task.sleep(nanoseconds: 2 * 1_000_000_000)
                
                guard let self = self else { return }
                await self.checkCPUUsageAndLevel()
            }
        }
        
        RJLogger.debug("Performance monitoring started")
    }
    
    @objc public func stopMonitoring() {
        if !isMonitoring { return }
        isMonitoring = false
        
        NotificationCenter.default.removeObserver(self)
        
        monitorTask?.cancel()
        monitorTask = nil
        
        if let source = memoryPressureSource {
            source.cancel()
            memoryPressureSource = nil
        }
        
        RJLogger.debug("Performance monitoring stopped")
    }
    
    @objc public func updatePerformanceLevel() {
        // Run on main actor or sync queue if needed, but since we modify property, let's keep it safe.
        // For simplicity, we can just trigger the check logic which handles logic.
        // But since this is public and might be called from any thread, let's use Task to be safe if called from ObjC
        Task { @MainActor in
            self.performUpdateCheck()
        }
    }
    
    // MARK: - Monitoring Logic
    
    private func setupMemoryPressureMonitoring() {
        let source = DispatchSource.makeMemoryPressureSource(eventMask: [.warning, .critical], queue: .main)
        
        source.setEventHandler { [weak self] in
             guard let self = self else { return }
             let event = source.data
             
             if event.contains(.critical) {
                 RJLogger.warning("CRITICAL memory pressure - pausing captures")
                 self.setLevel(.paused)
                 self.handleMemoryWarning()
             } else if event.contains(.warning) {
                 RJLogger.warning("Memory pressure warning - reducing captures")
                 self.setLevel(.minimal)
             }
        }
        
        source.resume()
        memoryPressureSource = source
    }
    
    @objc private func handleMemoryWarningNotification() {
        handleMemoryWarning()
    }
    
    @objc public func handleMemoryWarning() {
        RJLogger.warning("Memory warning received")
        setLevel(.minimal)
        delegate?.performanceManagerDidReceiveMemoryWarning?()
    }
    
    @objc private func handleThermalStateChange() {
        Task { @MainActor in performUpdateCheck() }
    }
    
    @objc private func handleBatteryStateChange() {
        Task { @MainActor in performUpdateCheck() }
    }
    
    @MainActor
    private func performUpdateCheck() {
        if !thermalThrottleEnabled && !batteryAwareEnabled {
            setLevel(.normal)
            return
        }
        
        var newLevel = RJPerformanceLevel.normal
        
        // Thermal State
        if thermalThrottleEnabled {
            let state = ProcessInfo.processInfo.thermalState
            switch state {
            case .critical:
                setLevel(.paused)
                return
            case .serious:
                setLevel(.minimal)
                return
            case .fair:
                newLevel = .reduced
            case .nominal:
                break
            @unknown default:
                break
            }
        }
        
        // Battery State
        if batteryAwareEnabled {
            let batteryState = UIDevice.current.batteryState
            let batteryLevel = UIDevice.current.batteryLevel
            
            let isCharging = (batteryState == .charging || batteryState == .full)
            if !isCharging && batteryLevel >= 0 && batteryLevel < lowBatteryThreshold {
                newLevel = maxLevel(newLevel, .reduced)
            }
        }
        
        // Memory Usage
        let usedMemory = currentMemoryUsage()
        if usedMemory > memoryWarningThresholdBytes {
             newLevel = maxLevel(newLevel, .reduced)
        }
        
        setLevel(newLevel)
    }
    
    private func setLevel(_ level: RJPerformanceLevel) {
        if currentLevel != level {
            currentLevel = level
        }
    }
    
    private func maxLevel(_ l1: RJPerformanceLevel, _ l2: RJPerformanceLevel) -> RJPerformanceLevel {
        return l1.rawValue > l2.rawValue ? l1 : l2
    }
    
    private func checkCPUUsageAndLevel() async {
        let cpu = currentCPUUsage()
        if cpu < 0 { return }
        
        if rollingCPUAverage == 0 {
            rollingCPUAverage = cpu
        } else {
            rollingCPUAverage = 0.3 * cpu + 0.7 * rollingCPUAverage
        }
        
        var suggestedLevel = RJPerformanceLevel.normal
        
        if rollingCPUAverage >= cpuCriticalThreshold {
            suggestedLevel = .minimal
            highCPUSampleCount += 1
        } else if rollingCPUAverage >= cpuHighThreshold {
            suggestedLevel = .reduced
            highCPUSampleCount += 1
        } else if rollingCPUAverage < cpuNormalThreshold {
            highCPUSampleCount = 0
            suggestedLevel = .normal
        } else {
            // Hysteresis zone
            if currentLevel.rawValue >= RJPerformanceLevel.reduced.rawValue {
                return
            }
            return
        }
        
        await MainActor.run {
            if suggestedLevel.rawValue > currentLevel.rawValue {
                if highCPUSampleCount >= hysteresisSamples {
                    RJLogger.info("CPU usage high (\(String(format: "%.1f", rollingCPUAverage))%), throttling to level \(suggestedLevel.rawValue)")
                    setLevel(suggestedLevel)
                }
            } else if suggestedLevel.rawValue < currentLevel.rawValue {
                if highCPUSampleCount == 0 {
                    RJLogger.info("CPU usage normalized (\(String(format: "%.1f", rollingCPUAverage))%), restoring level \(suggestedLevel.rawValue)")
                    setLevel(suggestedLevel)
                }
            }
        }
    }

    // MARK: - Metrics Calculation
    
    @objc public func currentMemoryUsage() -> UInt {
        var info = task_basic_info()
        var count = mach_msg_type_number_t(TASK_BASIC_INFO_COUNT)
        
        let kerr: kern_return_t = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                task_info(mach_task_self_, task_flavor_t(TASK_BASIC_INFO), $0, &count)
            }
        }
        
        if kerr == KERN_SUCCESS {
            return UInt(info.resident_size)
        } else {
            return 0
        }
    }
    
    @objc public func currentCPUUsage() -> Float {
        var totalUsage: Float = 0.0
        var threadList: thread_act_array_t?
        var threadCount: mach_msg_type_number_t = 0
        
        let kr = task_threads(mach_task_self_, &threadList, &threadCount)
        if kr != KERN_SUCCESS { return -1.0 }
        
        if let threadList = threadList {
            for i in 0..<Int(threadCount) {
                var threadInfo = thread_basic_info()
                var threadInfoCount = mach_msg_type_number_t(THREAD_INFO_MAX)
                
                let kerr: kern_return_t = withUnsafeMutablePointer(to: &threadInfo) {
                    $0.withMemoryRebound(to: integer_t.self, capacity: Int(threadInfoCount)) {
                        thread_info(threadList[i], thread_flavor_t(THREAD_BASIC_INFO), $0, &threadInfoCount)
                    }
                }
                
                if kerr == KERN_SUCCESS {
                    if (threadInfo.flags & TH_FLAGS_IDLE) == 0 {
                        totalUsage += Float(threadInfo.cpu_usage) / Float(TH_USAGE_SCALE) * 100.0
                    }
                }
            }
            
            vm_deallocate(mach_task_self_, vm_address_t(bitPattern: threadList), vm_size_t(threadCount * UInt32(MemoryLayout<thread_t>.stride)))
        }
        
        return totalUsage
    }
    
    @objc public var isCPUHigh: Bool {
        return rollingCPUAverage >= 60.0
    }
}
